flink-streaming-java/src/test/java/org/apache/flink/streaming/api/datastream/ReinterpretDataStreamAsKeyedStreamITCase.java

```diff
@@ -187,26 +208,71 @@ public void run(SourceContext<Tuple2<Integer, Integer>> out) throws Exception {
 			this.running = true;
 			try {
 				while (running) {
-					Integer key = din.readInt();
-					Integer val = din.readInt();
-					out.collect(new Tuple2<>(key, val));
+
+					checkFail();
+
+					synchronized (out.getCheckpointLock()) {
+						Integer key = din.readInt();
+						Integer val = din.readInt();
+						out.collect(new Tuple2<>(key, val));
+
+						position += 2 * Integer.BYTES;
+					}
 				}
 			} catch (EOFException ignore) {
+				while (!isRestored) {
```

Comment:
This while loop is a bit confusing for me.
Isn't `isRestored` set only once in `initializeState`?

Reply:
Yes, and that is ok. The loop is only there to prevent that the source will terminate before a successful checkpoint. After the failover, when `isRestore == true`, we allow that the source can terminate directly to finish the job.

Reply:
Yeah, I guess we can just check it once in the beginning as an `if`.